### PR TITLE
8337199: Add jcmd Thread.vthread_summary diagnostic command

### PR DESCRIPTION
**Pull request by AlanBateman**
Adds `jcmd <pid> Thread.vthread_summary` to print information to help diagnose issues with virtual threads. The output includes the virtual thread scheduler, the schedulers used to support timeouts, and the I/O pollers used to support virtual threads doing socket I/O, and a summary of the thread groupings.

This command has been in the loom repo since last year, it renamed a few times but we didn't find a better name. I'll paste in sample output in a PR comment.

As with a few other `jcmd` commands, this one is an upcall to a Java method that has access to the implementation classes. As with other `jcmd` commands, this one has to buffer the output so a limit it put on the number of thread groupings that is printed.

The jtreg failure handler is updated to run this jcmd, which can be used for test timeouts.

---------
### Progress
- [ ] Change must be properly reviewed (1 review required, with at least 1 [Reviewer](https://openjdk.org/bylaws#reviewer))
- [ ] Change requires CSR request [JDK-8339420](https://bugs.openjdk.org/browse/JDK-8339420) to be approved
- [x] Change must not contain extraneous whitespace
- [x] Commit message must refer to an issue

### Warnings
&nbsp;⚠️ Patch contains a binary file (src/utils/IdealGraphVisualizer/View/src/main/resources/com/sun/hotspot/igv/view/images/hideDuplicates.png)
&nbsp;⚠️ Patch contains a binary file (test/jdk/java/security/ProtectionDomain/AllPerm.jar)
&nbsp;⚠️ Patch contains a binary file (test/jdk/sun/security/provider/PolicyFile/TokenStore.keystore)
&nbsp;⚠️ Patch contains a binary file (test/jdk/sun/security/provider/PolicyFile/TrustedCert.keystore)
&nbsp;⚠️ Patch contains a binary file (test/jdk/sun/security/provider/PolicyFile/TrustedCert.keystore1)
&nbsp;⚠️ Patch contains a binary file (test/jdk/sun/security/provider/PolicyParser/ExtDirsA/a.jar)
&nbsp;⚠️ Patch contains a binary file (test/jdk/sun/security/provider/PolicyParser/ExtDirsB/b.jar)

### Issues
 * [JDK-8337199](https://bugs.openjdk.org/browse/JDK-8337199): Add jcmd Thread.vthread_summary diagnostic command (**Enhancement** - P4)
 * [JDK-8339420](https://bugs.openjdk.org/browse/JDK-8339420): Add jcmd Thread.vthread_summary diagnostic command (**CSR**)

### Reviewing
<details><summary>Using <code>git</code></summary>

Checkout this PR locally: \
`$ git fetch https://git.openjdk.org/jdk.git pull/20817/head:pull/20817` \
`$ git checkout pull/20817`

Update a local copy of the PR: \
`$ git checkout pull/20817` \
`$ git pull https://git.openjdk.org/jdk.git pull/20817/head`

</details>
<details><summary>Using Skara CLI tools</summary>

Checkout this PR locally: \
`$ git pr checkout 20817`

View PR using the GUI difftool: \
`$ git pr show -t 20817`

</details>
<details><summary>Using diff file</summary>

Download this PR as a diff file: \
<a href="https://git.openjdk.org/jdk/pull/20817.diff">https://git.openjdk.org/jdk/pull/20817.diff</a>

</details>

### CODE REVIEW

**Comment by AlanBateman** [comment]
Sample output on large Linux system. The I/O poller section varies by platform/configuration.
```
985461:
<root> [platform threads = 9, virtual threads = 512]
    +-- ForkJoinPool.commonPool/jdk.internal.vm.SharedThreadContainer@a295088 [platform threads = 0, virtual threads = 0]
    +-- java.util.concurrent.ScheduledThreadPoolExecutor@4fca772d [platform threads = 1, virtual threads = 0]
    +-- java.util.concurrent.ScheduledThreadPoolExecutor@27973e9b [platform threads = 1, virtual threads = 0]
    +-- java.util.concurrent.ThreadPerTaskExecutor@12f07c0b [platform threads = 0, virtual threads = 33]
    +-- ForkJoinPool-1/jdk.internal.vm.SharedThreadContainer@2861fe72 [platform threads = 32, virtual threads = 0]
    +-- java.util.concurrent.ScheduledThreadPoolExecutor@133314b [platform threads = 1, virtual threads = 0]
    +-- java.util.concurrent.ScheduledThreadPoolExecutor@1b2c6ec2 [platform threads = 1, virtual threads = 0]
    +-- java.util.concurrent.ScheduledThreadPoolExecutor@3941a79c [platform threads = 1, virtual threads = 0]
    +-- java.util.concurrent.ScheduledThreadPoolExecutor@7530d0a [platform threads = 1, virtual threads = 0]
    +-- java.util.concurrent.ScheduledThreadPoolExecutor@7cd84586 [platform threads = 1, virtual threads = 0]
    +-- java.util.concurrent.ScheduledThreadPoolExecutor@3d494fbf [platform threads = 1, virtual threads = 0]

Default virtual thread scheduler:
java.util.concurrent.ForkJoinPool@688c56e6[Running, parallelism = 32, size = 32, active = 32, running = 0, steals = 40639243, tasks = 133, submissions = 8]

Timeout schedulers:
[0] java.util.concurrent.ScheduledThreadPoolExecutor@27973e9b[Running, pool size = 1, active threads = 0, queued tasks = 43, completed tasks = 130]
[1] java.util.concurrent.ScheduledThreadPoolExecutor@7530d0a[Running, pool size = 1, active threads = 0, queued tasks = 44, completed tasks = 131]
[2] java.util.concurrent.ScheduledThreadPoolExecutor@3941a79c[Running, pool size = 1, active threads = 0, queued tasks = 17, completed tasks = 151]
[3] java.util.concurrent.ScheduledThreadPoolExecutor@4fca772d[Running, pool size = 1, active threads = 0, queued tasks = 27, completed tasks = 153]
[4] java.util.concurrent.ScheduledThreadPoolExecutor@3d494fbf[Running, pool size = 1, active threads = 0, queued tasks = 29, completed tasks = 127]
[5] java.util.concurrent.ScheduledThreadPoolExecutor@133314b[Running, pool size = 1, active threads = 0, queued tasks = 33, completed tasks = 118]
[6] java.util.concurrent.ScheduledThreadPoolExecutor@7cd84586[Running, pool size = 1, active threads = 0, queued tasks = 25, completed tasks = 118]
[7] java.util.concurrent.ScheduledThreadPoolExecutor@1b2c6ec2[Running, pool size = 1, active threads = 0, queued tasks = 26, completed tasks = 127]

Master I/O poller:
sun.nio.ch.EPollPoller@3d27b38a [registered = 4]

Read I/O pollers:
[0] sun.nio.ch.EPollPoller@1c9a93b9 [registered = 16]
[1] sun.nio.ch.EPollPoller@7a94c953 [registered = 16]
[2] sun.nio.ch.EPollPoller@238fa070 [registered = 4]
[3] sun.nio.ch.EPollPoller@50817b87 [registered = 16]
[4] sun.nio.ch.EPollPoller@18ce2d75 [registered = 3]
[5] sun.nio.ch.EPollPoller@44cca16e [registered = 14]
[6] sun.nio.ch.EPollPoller@1a1e15a7 [registered = 13]
[7] sun.nio.ch.EPollPoller@41181820 [registered = 14]
[8] sun.nio.ch.EPollPoller@28d420f3 [registered = 16]
[9] sun.nio.ch.EPollPoller@9b22356 [registered = 11]
[10] sun.nio.ch.EPollPoller@636bbcea [registered = 16]
[11] sun.nio.ch.EPollPoller@62b371bf [registered = 10]
[12] sun.nio.ch.EPollPoller@6b6916a2 [registered = 12]
[13] sun.nio.ch.EPollPoller@7359200a [registered = 2]
[14] sun.nio.ch.EPollPoller@4deb5e67 [registered = 15]
[15] sun.nio.ch.EPollPoller@342cf0ec [registered = 1]
[16] sun.nio.ch.EPollPoller@17d32ccd [registered = 16]
[17] sun.nio.ch.EPollPoller@2db0074d [registered = 5]
[18] sun.nio.ch.EPollPoller@3a3f7cfa [registered = 16]
[19] sun.nio.ch.EPollPoller@69c750c1 [registered = 10]
[20] sun.nio.ch.EPollPoller@67fcbc70 [registered = 14]
[21] sun.nio.ch.EPollPoller@4adb55ad [registered = 13]
[22] sun.nio.ch.EPollPoller@b4238a0 [registered = 13]
[23] sun.nio.ch.EPollPoller@58d03051 [registered = 15]
[24] sun.nio.ch.EPollPoller@5a4d1807 [registered = 7]
[25] sun.nio.ch.EPollPoller@4a2b48de [registered = 16]
[26] sun.nio.ch.EPollPoller@5036ee03 [registered = 6]
[27] sun.nio.ch.EPollPoller@9bd0eec [registered = 16]
[28] sun.nio.ch.EPollPoller@3b82b2fc [registered = 11]
[29] sun.nio.ch.EPollPoller@2b7f60ea [registered = 15]
[30] sun.nio.ch.EPollPoller@4cdec570 [registered = 16]
[31] sun.nio.ch.EPollPoller@2d756869 [registered = 12]

Write I/O pollers:
[0] sun.nio.ch.EPollPoller@536662da [registered = 0]
```

File: src/jdk.jcmd/share/man/jcmd.1 [hunk 1]
```diff
@@ -958,6 +958,13 @@ false)
 (BOOLEAN, false)
 .RE
 .TP
+\f[V]Thread.vthread_summary\f[R]
+Print virtual thread summary
+.RS
+.PP
+Impact: Low
+.RE
+.TP
```
Comment (plummercj) [inline src/jdk.jcmd/share/man/jcmd.1:967]: More detail on what the output contains?

File: src/java.base/share/classes/jdk/internal/vm/VThreadSummary.java [hunk 2]
```diff
@@ -0,0 +1,202 @@
...
+         * number of thread containers has already been printed.
+         * @param container the thread container
+         * @param depth the depth in the tree, for indentation purposes
+         * @return true if the thread container was printed, false if max already printed
+         */
+        private boolean printThreadContainer(ThreadContainer container, int depth) {
+            count++;
+            if (count > max) {
+                sb.append("<truncated ...>")
+                        .append(System.lineSeparator());
```
Comment (plummercj) [inline src/java.base/share/classes/jdk/internal/vm/VThreadSummary.java:118]: Would be nice to have a test case for truncation.

File: src/java.base/share/classes/jdk/internal/vm/VThreadSummary.java [hunk 3]
```diff
@@ -0,0 +1,202 @@
...
+                return false;
+            }
+
+            Map<Boolean, Long> threadCounts = container.threads()
+                    .collect(Collectors.partitioningBy(Thread::isVirtual, Collectors.counting()));
+            long platformThreadCount = threadCounts.get(Boolean.FALSE);
+            long virtualThreadCount = threadCounts.get(Boolean.TRUE);
+            if (depth > 0) {
+                int indent = depth * 4;
+                sb.append(" ".repeat(indent)).append("+-- ");
```
Comment (plummercj) [inline src/java.base/share/classes/jdk/internal/vm/VThreadSummary.java:128]: Would be nice to have an indentation test case.

**Comment by AlanBateman** [comment]
Larry is taking this so I will close this PR.

**Pull request closed without merging**